dataprovider/bolt.go

```diff
@@ -352,6 +352,7 @@ func (p BoltProvider) updateUser(user User) error {
 		user.UsedQuotaSize = oldUser.UsedQuotaSize
 		user.UsedQuotaFiles = oldUser.UsedQuotaFiles
 		user.LastLogin = oldUser.LastLogin
+		user.ID = oldUser.ID
```

Comment:
Is this required? If so we should add it to the memory provider too. This case should be checked [here](https://github.com/drakkan/sftpgo/blob/master/httpd/api_user.go#L141) already. Is there some failing test case?

Reply:
This was causing most of the modified InvalidFs tests to fail because they do this:

1. Add user
2. Update user (based on add user response)
3. Delete user (based on add user response)

The flow for the Bolt data provider was:
1. Add user is fine, returns a non-zero user ID
2. Update user is called with a valid user, but due to this issue, stores a user with ID 0 at the correct user ID's key.
3. Delete user 500s because the HTTP handler uses a user ID path parameter, looks the user up, and then passes it to Bolt's `deleteUser` which does this:
```
        userIDAsBytes := itob(user.ID)
        userName := idxBucket.Get(userIDAsBytes)
```

At a glance, the memory provider doesn't suffer from this because it re-uses the passed-in user (preserving the user ID by default) where Bolt makes a new user struct instance based on the previous one.

Reply:
This likely happen because inside the tests you update the user this way:

```
err = dataprovider.UpdateUser(u)
```

without using the (public) REST API. It should not happen (or fail before) if you use `httpd.UpdateUser(....`. I'll do some tests myself later today, thank you